service/src/main/java/bio/terra/axonserver/utils/CloudStorageUtils.java

```diff
@@ -70,15 +77,37 @@ public static byte[] getBucketObject(
             .getService()
             .reader(blobId)) {
 
+      long startByteIdx =
+          Optional.ofNullable(byteRange).isPresent() ? byteRange.getRangeStart(Long.MAX_VALUE) : 0;
+      long endByteIdx =
+          Optional.ofNullable(byteRange).isPresent()
+              ? byteRange.getRangeEnd(Long.MAX_VALUE)
+              : Long.MAX_VALUE - 1;
+      long bytesToRead = endByteIdx - startByteIdx;
+
       BoundedByteArrayOutputStream outputStream = new BoundedByteArrayOutputStream(MAX_OBJECT_SIZE);
-      ByteBuffer bytes = ByteBuffer.allocate(MAX_BUFFER_SIZE);
-      while (reader.read(bytes) > 0) {
-        bytes.flip();
-        outputStream.write(bytes.array(), 0, bytes.limit());
-        bytes.clear();
+      ByteBuffer buffer = ByteBuffer.allocate(MAX_BUFFER_SIZE);
+
+      long totalBytesRead = 0;
+      int bytesRead;
+      reader.seek(startByteIdx);
+      while (totalBytesRead < bytesToRead && reader.read(buffer) > 0) {
+        bytesRead = buffer.position();
+
+        // write bytes in buffer to output stream
+        // limit the number of bytes written to the number of bytes left to read
+        buffer.flip();
+        int bufferLimit = (int) Math.min(buffer.limit(), bytesToRead - totalBytesRead);
```

Comment:
There's an edge case where if a provided byte range is less than the size of the buffer and the reader still has more bytes, the code will write bytes beyond the request range into the output stream. The reason why it seems to still work though is because spring has built in support for partial ranges when the return type is Resource.

I added a fix to only write the remaining amount of bytes requested and not more. Also had to change the return type of the controller to byte[] so [spring doesn't mess with the response](https://github.com/spring-projects/spring-framework/blob/1984cfe9d79e8a34955a393ef9d71a2a3d11a87c/spring-webmvc/src/main/java/org/springframework/web/servlet/mvc/method/annotation/AbstractMessageConverterMethodProcessor.java#L196-L213) instead correctly returns the full byte array.